module/zfs/spa.c

```diff
@@ -3326,6 +3338,106 @@ spa_start_livelist_condensing_thread(spa_t *spa)
 	    spa_livelist_condense_cb, spa, minclsyspri);
 }
 
+static int
+spa_rrd_write(spa_t *spa, dmu_tx_t *tx, rrd_t *rrd, const char *what,
+    boolean_t insert)
+{
+	static uint8_t buf[sizeof (*rrd)];
+
+	rrd_pack(rrd, buf);
+
+	if (insert) {
+		return (zap_add(spa_meta_objset(spa), DMU_POOL_DIRECTORY_OBJECT,
+		    what, 1, sizeof (*rrd), buf, tx));
+	}
+
+	return (zap_update(spa_meta_objset(spa),
+	    DMU_POOL_DIRECTORY_OBJECT, what, 1, sizeof (*rrd), buf,
+	    tx));
+}
+
+static void
+spa_sync_time_logger(spa_t *spa, uint64_t txg)
+{
+	uint64_t curtime;
+	boolean_t insert;
+	dmu_tx_t *tx;
+
+	if (!spa_writeable(spa)) {
+		return;
+	}
+	if (txg == spa->spa_last_noted_txg) {
+		return;
+	}
+	curtime = gethrestime_sec();
+	if (curtime < spa->spa_last_noted_txg_time + spa_note_txg_time) {
+		return;
+	}
+
+	spa->spa_last_noted_txg_time = curtime;
+	spa->spa_last_noted_txg = txg;
+
+	mutex_enter(&spa->spa_txg_log_time_lock);
+	dbrrd_add(&spa->spa_txg_log_time, curtime, txg);
+
+	if (curtime < spa->spa_last_flush_txg_time + spa_flush_txg_time) {
+		goto out;
+	}
+	spa->spa_last_flush_txg_time = curtime;
+
+	insert = (zap_contains(spa_meta_objset(spa), DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_TXG_LOG_TIME_MINUTES) == ENOENT);
+
+	tx = dmu_tx_create_assigned(spa_get_dsl(spa), txg);
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_minutes,
+	    DMU_POOL_TXG_LOG_TIME_MINUTES, insert));
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_days,
+	    DMU_POOL_TXG_LOG_TIME_DAYS, insert));
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_days,
+	    DMU_POOL_TXG_LOG_TIME_MONTHS, insert));
```

Comment:
Would it be simpler to store a single unix timestamp for the log time, rather than discrete  minutes/days/months?

Also, you're storing `dbr_days` into `DMU_POOL_TXG_LOG_TIME_MONTHS`.  Typo?

Reply:
Good catch!

The idea was to have different resolutions, so as long as the ZFS pool runs, we keep fewer historic TXGs.
We are also limited by the size of the ZAP entry.

Reply:
I'm concerned you're going to run into timezone issues if you don't use Unix timestamps.  If you write a timestamp at 10am EST in New York, and then later import the pool in San Francisco and write a timestamp at 10am PST, will there be two timestamps at "10am"?

Reply:
Oh... You are right me goal was to store time in UTC, the `zpool scrub` converts this time to be UTC.
I was sure the `gethrtime` return a UTC time stamp, but you are right it doesn't.
I have to fix that.

Reply:
Ok I hace double check, and I think that the implementation is correct.
The [gethrestime_sec(void)](https://github.com/openzfs/zfs/blob/21205f6488ade598427cc525b349be75a69d69bd/include/os/linux/spl/sys/time.h#L77) function uses [ktime_get_coarse_real_ts64](https://docs.kernel.org/core-api/timekeeping.html) which from my understanding returns a UTC time.